### PR TITLE
 Laboratory work #2, Ilyusheva Polina - 19FPL2

### PR DESCRIPTION
**Pull request by psilyusheva**

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -34,7 +47,24 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
...
+            return []
+    if not scd_sent_not_t:
+        if not all(isinstance(word, str) for word in second_sentence_tokens):
+            return []
+    if frst_sent_not_t or scd_sent_not_t:
+        return []
+    lcs_matrix = create_zero_matrix(len(first_sentence_tokens), len(second_sentence_tokens))
+    for index_1, element_1 in enumerate(first_sentence_tokens):
+        for index_2, element_2 in enumerate(second_sentence_tokens):
+            if element_1 == element_2 and index_1 == index_2:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: why index_1==index_2?
what happens if index_1==0 or index_2==0?

File: lab_2/main.py [hunk 2]
```diff
@@ -57,7 +105,41 @@ def find_lcs(first_sentence_tokens: tuple, second_sentence_tokens: tuple, lcs_ma
...
+    lcs = []
+    ind_r = len(first_sentence_tokens) - 1
+    ind_c = len(second_sentence_tokens) - 1
+
+    while ind_r >= 0 and ind_c >= 0:
+        if first_sentence_tokens[ind_r] == second_sentence_tokens[ind_c]:
+            lcs.append(first_sentence_tokens[ind_r])
+            ind_r -= 1
+            ind_c -= 1
+        elif ind_r > ind_c:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: why do you compare indexes, not elements?

File: lab_2/main.py [hunk 3]
```diff
@@ -81,7 +177,40 @@ def calculate_text_plagiarism_score(original_text_tokens: tuple, suspicious_text
...
+    sus_text_not_tuple = not isinstance(suspicious_text_tokens, tuple)
+    plagiarism_not_float = not isinstance(plagiarism_threshold, float)
+    if not orig_txt_not_tuple:
+        for i in range(len(original_text_tokens)):
+            if type(original_text_tokens[i]) == tuple:
+                if not all(isinstance(word, str) for word in original_text_tokens[i]):
+                    return -1.0
+            else:
+                return -1.0
+    if not sus_text_not_tuple:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: is there a way to join this check and the previous one?

File: lab_2/main.py [hunk 4]
```diff
@@ -92,7 +221,52 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+        if not all(isinstance(word, str) for word in lcs):
+            return ()
+
+    org_dif_wds_ind = [original_sentence_tokens.index(word) for word in original_sentence_tokens if word not in lcs]
+    sus_dif_wds_ind = [suspicious_sentence_tokens.index(word) for word in suspicious_sentence_tokens if word not in lcs]
+    changed_ind_list = []
+    changed_ind_list_1 = []
+    changed_ind_list_2 = []
+
+    def get_changed_list(changed_ind_list, dif_wds_ind):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: find a way to do this without a function (try to iterate over two tuples )

File: lab_2/main.py [hunk 5]
```diff
@@ -282,7 +282,39 @@ def accumulate_diff_stats(original_text_tokens: tuple, suspicious_text_tokens: t
...
+                                                      suspicious_text_tokens,
+                                                      plagiarism_threshold)
+    sentence_plagiarism, sentence_lcs_length, difference_indexes = [], [], []
+
+    ind = 0
+    while ind != len(original_text_tokens):
+        sentence_lcs_length_i = find_lcs_length(original_text_tokens[ind],
+                                                suspicious_text_tokens[ind],
+                                                plagiarism_threshold)
+        sentence_lcs_length.append(sentence_lcs_length_i)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: can you work directly with the dictionary `statistics`?

File: lab_2/main.py [hunk 6]
```diff
@@ -49,18 +51,16 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
...
+            if not all(isinstance(word, str) for word in function_parameter):
+                return []
     if frst_sent_not_t or scd_sent_not_t:
         return []
     lcs_matrix = create_zero_matrix(len(first_sentence_tokens), len(second_sentence_tokens))
     for index_1, element_1 in enumerate(first_sentence_tokens):
         for index_2, element_2 in enumerate(second_sentence_tokens):
-            if element_1 == element_2 and index_1 == index_2:
+            if element_1 == element_2:
                 lcs_matrix[index_1][index_2] = lcs_matrix[index_1 - 1][index_2 - 1] + 1
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:64]: still, I have a question, what happens if index_1 == 0 or index_2 == 0? 
what is `lcs_matrix[index_1 - 1][index_2 - 1]`  then?

File: lab_2/main.py [hunk 7]
```diff
@@ -91,6 +91,8 @@ def find_lcs_length(first_sentence_tokens: tuple, second_sentence_tokens: tuple,
         return -1
     if not first_sentence_tokens or not second_sentence_tokens:
         return 0
+    if len(first_sentence_tokens) > len(second_sentence_tokens):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:94]: do you need this check?

File: lab_2/main.py [hunk 8]
```diff
@@ -81,7 +183,36 @@ def calculate_text_plagiarism_score(original_text_tokens: tuple, suspicious_text
...
+                    return -1.0
+
+    if orig_txt_not_tuple or sus_text_not_tuple \
+            or plagiarism_not_float or 0 > plagiarism_threshold > 1:
+        return -1.0
+
+    original_list = list(original_text_tokens)
+    if len(original_text_tokens) < len(suspicious_text_tokens):
+        original_list.append(()*(len(suspicious_text_tokens) - len(original_text_tokens)))
+    elif len(suspicious_text_tokens) < len(original_text_tokens):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:206]: do you need this?

File: lab_2/main.py [hunk 9]
```diff
@@ -92,7 +223,39 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+                return ()
+
+    org_dif_wds_ind = [original_sentence_tokens.index(word) for word in original_sentence_tokens if word not in lcs]
+    sus_dif_wds_ind = [suspicious_sentence_tokens.index(word) for word in suspicious_sentence_tokens if word not in lcs]
+    changed_ind_list = []
+    changed_ind_list_orig = []
+    changed_ind_list_sus = []
+
+    for data in ([org_dif_wds_ind, changed_ind_list_orig], [sus_dif_wds_ind, changed_ind_list_sus]):
+        for ind, dif_wds_ind in enumerate(data[0]):
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:244]: be ready to explain this part

File: lab_2/main.py [hunk 10]
```diff
@@ -108,8 +271,44 @@ def accumulate_diff_stats(original_text_tokens: tuple, suspicious_text_tokens: t
...
+        lcs_matrix_i = fill_lcs_matrix(original_text_tokens[ind], suspicious_text_tokens[ind])
+        lcs_i = find_lcs(original_text_tokens[ind], suspicious_text_tokens[ind], lcs_matrix_i)
+        difference_indexes_i = find_diff_in_sentence(original_text_tokens[ind],
+                                                     suspicious_text_tokens[ind],
+                                                     lcs_i)
+        statistics['difference_indexes'].append(difference_indexes_i)
+        ind += 1
+    return statistics
+
+    # statistics = {'text_plagiarism': text_plagiarism,
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: remove comments

File: lab_2/main.py [hunk 11]
```diff
@@ -119,8 +318,40 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
+    ind = 0
+    report = ""
+    while ind != len(original_text_tokens):
+        different_indexes = report_data['difference_indexes'][ind]
+        report += "- "
+        for index, word in enumerate(original_text_tokens[ind]):
+            if index in different_indexes[0]:
+                report += "| "
+            report += "{} ".format(word)
+        # if len(original_text_tokens[ind]) in different_indexes[0]:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: no comments

**Review by ValeriyaKuznetsova — COMMENTED** [review]
correct the issues

everything else seems fine

File: lab_2/main.py (hunk 7 above)
Comment (psilyusheva) [reply lab_2/main.py:94]: i think i should have this so the code can handle different sized token inputs

File: lab_2/main.py (hunk 8 above)
Comment (psilyusheva) [reply lab_2/main.py:206]: same here, i do this so the code can handle different sized inputs

File: lab_2/main.py (hunk 7 above)
Comment (ValeriyaKuznetsova) [reply lab_2/main.py:94]: Ok

File: lab_2/main.py (hunk 6 above)
Comment (psilyusheva) [reply lab_2/main.py:64]: so, the lcs_matrix[index_1 - 1][index_2 - 1] is the last element of our matrix (which is zero matrix), which is 0

File: lab_2/main.py (hunk 9 above)
Comment (psilyusheva) [reply lab_2/main.py:244]: ok!

**Pull request closed without merging**